cachito/workers/pkg_managers/npm.py

```diff
@@ -0,0 +1,120 @@
+        # since they are properly cached in the parent dependency in Nexus
+        if not info.get("bundled", False) and "resolved" not in info:
+            log.error("The dependency %r is not from the npm registry", info)
+            raise CachitoError(
+                "The lock file contains a dependency not from the npm registry. "
+                "This is not yet supported."
+            )
+
+        dep = {
+            "bundled": info.get("bundled", False),
+            "dev": info.get("dev", False),
+            "name": name,
+            "type": "npm",
+            "version": info["version"],
+        }
+        name_to_deps.setdefault(name, [])
+        for d in name_to_deps[name]:
+            if d["version"] == dep["version"]:
+                # If a duplicate version was found but this one isn't bundled, then mark the
+                # dependency as not bundled so it's included individually in the deps directory
+                if not dep["bundled"]:
+                    d["bundled"] = False
+                # If a duplicate version was found but this one isn't a dev dependency, then mark
+                # the dependency as not dev
+                if not dep["dev"]:
+                    d["dev"] = False
+                break
+        else:
+            name_to_deps[name].append(dep)
+
+        if "dependencies" in info:
+            _get_deps(info["dependencies"], name_to_deps)
+
+
+def get_package_and_deps(package_lock_path):
+    """
+    Get the main package and dependencies based on the lock file.
+
+    :param str package_lock_path: the path to the lock file
+    :return: the dictionary of the main package and the list of dependencies
+    :rtype: (dict, list)
+    """
+    with open(package_lock_path, "r") as f:
+        package_lock = json.load(f)
+
+    package = {"name": package_lock["name"], "type": "npm", "version": package_lock["version"]}
+
+    name_to_deps = {}
+    _get_deps(package_lock.get("dependencies", {}), name_to_deps)
+    # Convert the name_to_deps mapping to a list now that it's fully populated
+    deps = [dep_info for deps_info in name_to_deps.values() for dep_info in deps_info]
+
+    return package, deps
+
+
+def resolve_npm(app_source_path, request):
+    """
+    Resolve and fetch npm dependencies for the given app source archive.
+
+    :param str app_source_path: the full path to the application source code
+    :param dict request: the Cachito request this is for
+    :return: a tuple of the npm package itself and the list of dictionaries representing the
+        dependencies
+    :rtype: (dict, list)
+    :raises CachitoError: if fetching the dependencies fails
+    """
+    # npm-shrinkwrap.json and package-lock.json share the same format but serve slightly
+    # different purposes. See the following documentation for more information:
+    # https://docs.npmjs.com/files/package-lock.json.
+    for lock_file in ("npm-shrinkwrap.json", "package-lock.json"):
```

Comment:
If neither is found, we'd get an obscure `NameError`. Can we make the error nicer?

Reply:
Sure. I was relying on the check in the calling function of `fetch_npm_source`, but you're right, it's best to be explicit.